### PR TITLE
Fixed resizer initial percentage values for side images

### PR DESCRIPTION
**Pull request by mlewand**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Initial resize of a side image with no width predefined now gives correct percentage values.

---

### Additional information

Related issue ckeditor/ckeditor5-image#306.

### CODE REVIEW

**Comment by Reinmar** [comment]
I'd like to look at this too, so I removed the R+ label for now.

File: src/widgetresize/resizerstate.js [hunk 1]
```diff
@@ -146,6 +141,23 @@ export default class ResizeState {
 		this.proposedHandleHostWidth = newSize.handleHostWidth;
 		this.proposedHandleHostHeight = newSize.handleHostHeight;
 	}
+
+	/**
+	 * Calculates a relative width of a `domResizeHost` compared to it's parent in percents.
+	 *
+	 * @protected
```
Comment (Reinmar) [inline src/widgetresize/resizerstate.js]: No one else uses this method so it should be private.

```suggestion
	 * @private
```

File: src/widgetresize/resizerstate.js [hunk 2]
```diff
@@ -146,6 +141,23 @@ export default class ResizeState {
...
+	 * @returns {Number}
+	 */
+	_getParentPercentage( domResizeHost ) {
+		const rect = new Rect( domResizeHost );
+		const domResizeHostParent = domResizeHost.parentElement;
+		// Need to use computed style as it properly excludes parent's paddings from the returned value.
+		const parentWidth = parseFloat( domResizeHostParent.ownerDocument.defaultView.getComputedStyle( domResizeHostParent ).width );
+
+		// Round to two digits in fraction.
+		return Math.round( rect.width / parentWidth * 10000 ) / 100;
```
Comment (Reinmar) [inline src/widgetresize/resizerstate.js]: Rounding here is unnecessary. We only need to round saved and displayed values. The internal state can use full resolution.

File: src/widgetresize/resizerstate.js [hunk 3]
```diff
@@ -146,6 +141,23 @@ export default class ResizeState {
...
 	}
+
+	/**
+	 * Calculates a relative width of a `domResizeHost` compared to it's parent in percents.
+	 *
+	 * @protected
+	 * @param {HTMLElement} domResizeHost
+	 * @returns {Number}
+	 */
+	_getParentPercentage( domResizeHost ) {
```
Comment (Reinmar) [inline src/widgetresize/resizerstate.js]:
```suggestion
	_calculateHostPercentageWidth( domResizeHost ) {
```

Comment (Reinmar) [reply src/widgetresize/resizerstate.js]: Plus, this function has nothing to do with this class – you can see that it does not access `this` even once. Hence, it's just a helper.

File: src/widgetresize/resizerstate.js [hunk 4]
```diff
@@ -146,6 +141,23 @@ export default class ResizeState {
...
+		const rect = new Rect( domResizeHost );
+		const domResizeHostParent = domResizeHost.parentElement;
```
Comment (Reinmar) [inline src/widgetresize/resizerstate.js]: Isn't this a bit naive? The image still blinks heavily on the first resize event (which I know is a separate issue), so it's a bit hard to tell how well we calculate the width. But if you're also aware that this is a naive solution, then it'd be good to comment this in the code.

Comment (mlewand) [reply src/widgetresize/resizerstate.js]: > The image still blinks heavily on the first resize event (which I know is a separate issue), so it's a bit hard to tell how well we calculate the width.

I'm pretty confident that this fix resolves the problem of this issue. You can see that the values are now calculated correctly. As you mentioned, the only problem is this full-width flash which will be fixed in a separated issue.

> Isn't this a bit naive? The image still blinks heavily on the first resize event (which I know is a separate issue), so it's a bit hard to tell how well we calculate the width.

TBH, I don't see anything too naive here. Max element's width is simply constrained by the parent, which is a `div`, `td`, `th` or `blockquote` etc.

If we want to make it more complex, then we can go with [`HTMLElement.offsetParent`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetParent).

The benefit of this solution is that it will work well with an image wrapped in a transparent content model container &mdash; I'm thinking mostly about `<a>` element, if we ever allow to wrap multiple blocks in a single `<a>` element. Without this, the current proposal would use this `<a>` directly and use `auto` as a width value, which will cause an issue.

The problem with `offsetParent` is that it comes with more cases where a `null` is returned and it would be nice to handle it somewhat gracefully. This would require `elem.offsetParent || <fallbackElement>` here, which would in turn require us to write weird and unrealistic test cases for code coverage.

All in all, simple parent usage works well for all the cases that we have today, so I see no reason to further complicate it for the time being.

**Comment by mlewand** [comment]
* I also reused resize host's Rect instnace in the 3ec9957.
* In b6a468c I have also aligned module's local JSDocs with the rest of the codebase.

File: src/widgetresize/resizerstate.js [hunk 5]
```diff
@@ -126,11 +126,8 @@ export default class ResizeState {
 
 		const widthStyle = domResizeHost.style.width;
 
-		if ( widthStyle && widthStyle.match( /^\d+\.?\d*%$/ ) ) {
-			this.originalWidthPercents = parseFloat( widthStyle );
-		} else {
-			this.originalWidthPercents = this._getParentPercentage( domResizeHost );
-		}
+		this.originalWidthPercents = widthStyle && widthStyle.match( /^\d+\.?\d*%$/ ) ?
```
Comment (Reinmar) [inline src/widgetresize/resizerstate.js]: Why this change? It makes the code harder to read.

**Comment by Reinmar** [comment]
OK, just a minor issue in the code style, and some minor code issues in https://github.com/ckeditor/ckeditor5-image/pull/308. Other than that, both changes look good to me. Once you resolve them and someone checks that the patch still works fine, you can merge this.

File: src/widgetresize/resizerstate.js (hunk 5 above)
Comment (mlewand) [reply src/widgetresize/resizerstate.js]: Is it really that harder to read? See the code in full context. Note that it now replaces:

```js
if ( widthStyle && widthStyle.match( /^\d+\.?\d*%$/ ) ) {
	this.originalWidthPercents = parseFloat( widthStyle );
} else {
	this.originalWidthPercents = calculateHostPercentageWidth( domResizeHost );
}
```

Rather than the previous form.

Comment (Reinmar) [reply src/widgetresize/resizerstate.js]: Yes, it is. Because if-else is *the* way to write a condition. If you see it, you know there's a condition. Your current code must be first parsed by the reader to understand what's happening.
